clarity/src/vm/types/serialization.rs

```diff
@@ -287,6 +291,132 @@ macro_rules! check_match {
+    /// Clarity value of this type. The returned length *may* not fit
+    /// in in a Clarity buffer! For example, the maximum serialized
+    /// size of a `(buff 1024*1024)` is `1+1024*1024` because of the
+    /// type prefix byte. However, that is 1 byte larger than the maximum
+    /// buffer size in Clarity.
+    pub fn max_serialized_size(&self) -> Result<u32, CheckErrors> {
+        let type_prefix_size = 1;
+
+        let max_output_size = match self {
+            TypeSignature::NoType => {
+                // A `NoType` should *never* actually be evaluated
+                // (`NoType` corresponds to the Some branch of a
+                // `none` that is never matched with a corresponding
+                // `some` or similar with `result` types).  So, when
+                // serializing an object with a `NoType`, the other
+                // branch should always be used.
+                return Err(CheckErrors::CouldNotDetermineSerializationType);
+            }
+            TypeSignature::IntType => 16,
+            TypeSignature::UIntType => 16,
+            TypeSignature::BoolType => 0,
+            TypeSignature::SequenceType(SequenceSubtype::ListType(list_type)) => {
+                // u32 length as big-endian bytes
+                let list_length_encode = 4;
+                list_type
+                    .get_max_len()
+                    .checked_mul(list_type.get_list_item_type().max_serialized_size()?)
+                    .and_then(|x| x.checked_add(list_length_encode))
+                    .ok_or_else(|| CheckErrors::ValueTooLarge)?
+            }
+            TypeSignature::SequenceType(SequenceSubtype::BufferType(buff_length)) => {
+                // u32 length as big-endian bytes
+                let buff_length_encode = 4;
+                u32::from(buff_length)
+                    .checked_add(buff_length_encode)
+                    .ok_or_else(|| CheckErrors::ValueTooLarge)?
+            }
+            TypeSignature::SequenceType(SequenceSubtype::StringType(StringSubtype::ASCII(
+                length,
+            ))) => {
+                // u32 length as big-endian bytes
+                let str_length_encode = 4;
+                // ascii is 1-byte per character
+                u32::from(length)
+                    .checked_add(str_length_encode)
+                    .ok_or_else(|| CheckErrors::ValueTooLarge)?
+            }
+            TypeSignature::SequenceType(SequenceSubtype::StringType(StringSubtype::UTF8(
+                length,
+            ))) => {
+                // u32 length as big-endian bytes
+                let str_length_encode = 4;
+                // utf-8 is maximum 4 bytes per codepoint (which is the length)
+                u32::from(length)
+                    .checked_mul(4)
+                    .and_then(|x| x.checked_add(str_length_encode))
+                    .ok_or_else(|| CheckErrors::ValueTooLarge)?
+            }
+            TypeSignature::PrincipalType => {
```

Comment:
It's a pity that the `PrincipalType`'s maximum size calculation must assume a maximally-long contract principal, instead of distinguishing between standard and contract principals.  But given our time tables, I'm assuming that this is the right thing to do for now.

Reply:
I think this is unavoidable as long as the same type encompasses both contract and standard principals, and I'd be pretty resistant to making new types `standard-principal` and `contract-principal`, because I think we should encourage contracts to treat all principals the same.